### PR TITLE
Supporting the S7 Elephant Rev2 PCB

### PR DESCRIPTION
**Pull request by MudkipMao**
S7 Elephant had a GB in late 2019 that is now shipping to customers. It features a new PCB with backlight compatibility. This change aims to bring QMK compatibility as well as allowing for VIA compatibility in the future.

## Description

I refactored the whole folder into a rev1 and rev2 scheme since the MCU pinouts are very different between the two PCB's. I tried to leave rev1 intact but do not have a PCB to test with. Rev2 is tested on my own PCB. Additionally, I have laid out the groundwork for VIA support which I am planning to finish this coming week.

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [x] Enhancement/optimization
- [x] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## What needs work
* Finishing VIA support
* Adding more layouts, rev1 has many more layouts to make it easier for people to customize. Currently rev2 only has LAYOUT_all and LAYOUT_ansi

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by MudkipMao** [comment]
Also, another question. For the VIA firmware I built for the rev2 PCB, the firmware only has 450 bytes free. Is this cutting it too close? I can disable CONSOLE_ENABLE and COMMAND_ENABLE to free up more space

File: keyboards/s7_elephant/rev2/keymaps/via/rules.mk [hunk 1]
```diff
@@ -0,0 +1,4 @@
+# Build options
+
+VIA_ENABLE = yes         # Enable usage of VIA configurator
```
Comment (mechmerlin) [inline keyboards/s7_elephant/rev2/keymaps/via/rules.mk:3]: LTO_ENABLE = yes

File: keyboards/s7_elephant/rev2/rules.mk [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+BLUETOOTH_ENABLE = no        # Enable Bluetooth with the Adafruit EZ-Key HID
+AUDIO_ENABLE = no            # Audio output on port C6
+FAUXCLICKY_ENABLE = no       # Use buzzer to emulate clicky switches
+HD44780_ENABLE = no          # Enable support for HD44780 based LCDs
+
+# Adding these flags to reduce image size because VIA firmware is too large and default is cutting it close
+EXTRAFLAGS += -flto # Don't remove this without also removing NO_ACTION_MACRO and NO_ACTION_FUNCTION in config.h
+
+# generated by KBFirmware JSON to QMK Parser
+# https://noroadsleft.github.io/kbf_qmk_converter/
```
Comment (fauxpark) [inline keyboards/s7_elephant/rev2/rules.mk:29]:
```suggestion
LTO_ENABLE = yes
```

File: keyboards/s7_elephant/rev2/rules.mk [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+#   change yes to no to disable
+#
+BOOTMAGIC_ENABLE = lite      # Virtual DIP switch configuration
+MOUSEKEY_ENABLE = yes        # Mouse keys
+EXTRAKEY_ENABLE = yes        # Audio control and System control
+
+# Consider setting these 2 to "no" if firmware goes over size. Disabling these will save you ~8kB
+CONSOLE_ENABLE = yes         # Console for debug
+COMMAND_ENABLE = yes         # Commands for debug and configuration
+
```
Comment (fauxpark) [inline keyboards/s7_elephant/rev2/rules.mk:11]:
```suggestion
CONSOLE_ENABLE = no          # Console for debug
COMMAND_ENABLE = no          # Commands for debug and configuration
```

File: keyboards/s7_elephant/rev2/readme.md [hunk 4]
```diff
@@ -0,0 +1,17 @@
+# S7 Elephant
+
+![S7 Elephant](https://i.imgur.com/CaHxKBJ.jpg)
+
+A high-end 70% custom keyboard designed and produced by [jacky](https://geekhack.org/index.php?action=profile;u=63864) from Geekhack.
+
+Keyboard Maintainer: QMK Community  
+Hardware Supported: S7 Elephant  
+Hardware Availability: [The GB has ended with little chance to run again.](https://geekhack.org/index.php?topic=96434.0)
```
Comment (fauxpark) [inline keyboards/s7_elephant/rev2/readme.md:9]:
```suggestion
* Keyboard Maintainer: QMK Community
* Hardware Supported: S7 Elephant
* Hardware Availability: [The GB has ended with little chance to run again.](https://geekhack.org/index.php?topic=96434.0)
```

File: keyboards/s7_elephant/rev2/readme.md [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+
+A high-end 70% custom keyboard designed and produced by [jacky](https://geekhack.org/index.php?action=profile;u=63864) from Geekhack.
+
+Keyboard Maintainer: QMK Community  
+Hardware Supported: S7 Elephant  
+Hardware Availability: [The GB has ended with little chance to run again.](https://geekhack.org/index.php?topic=96434.0)  
+
+Make example for this keyboard (after setting up your build environment):
+
+    make s7_elephant/rev2:default:dfu
```
Comment (fauxpark) [inline keyboards/s7_elephant/rev2/readme.md:13]:
```suggestion
    make s7_elephant/rev2:default
```

File: keyboards/s7_elephant/rev1/readme.md [hunk 6]
```diff
@@ -0,0 +1,17 @@
+# S7 Elephant
+
+![S7 Elephant](https://i.imgur.com/CaHxKBJ.jpg)
+
+A high-end 70% custom keyboard designed and produced by [jacky](https://geekhack.org/index.php?action=profile;u=63864) from Geekhack.
+
+Keyboard Maintainer: QMK Community  
+Hardware Supported: S7 Elephant  
+Hardware Availability: [The GB has ended with little chance to run again.](https://geekhack.org/index.php?topic=96434.0)
```
Comment (fauxpark) [inline keyboards/s7_elephant/rev1/readme.md:9]:
```suggestion
* Keyboard Maintainer: QMK Community
* Hardware Supported: S7 Elephant
* Hardware Availability: [The GB has ended with little chance to run again.](https://geekhack.org/index.php?topic=96434.0)
```

File: keyboards/s7_elephant/rev1/readme.md [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+
+A high-end 70% custom keyboard designed and produced by [jacky](https://geekhack.org/index.php?action=profile;u=63864) from Geekhack.
+
+Keyboard Maintainer: QMK Community  
+Hardware Supported: S7 Elephant  
+Hardware Availability: [The GB has ended with little chance to run again.](https://geekhack.org/index.php?topic=96434.0)  
+
+Make example for this keyboard (after setting up your build environment):
+
+    make s7_elephant/rev1:default:dfu
```
Comment (fauxpark) [inline keyboards/s7_elephant/rev1/readme.md:13]:
```suggestion
    make s7_elephant/rev1:default
```

File: keyboards/s7_elephant/readme.md [hunk 8]
```diff
@@ -5,12 +5,17 @@
...
 
 Make example for this keyboard (after setting up your build environment):
 
-    make s7_elephant:default:dfu
+    make s7_elephant/rev2:default
+
+Install examples:
+
+    make s7_elephant/rev1:default:dfu         # For rev1 (2018 run)
+    make s7_elephant/rev2:default:dfu         # For rev1 (2019 run)
```
Comment (fauxpark) [inline keyboards/s7_elephant/readme.md:18]:
```suggestion
    make s7_elephant/rev1:default # 2018 run
    make s7_elephant/rev2:default # 2019 run
```

File: keyboards/s7_elephant/readme.md [hunk 9]
```diff
@@ -5,12 +5,17 @@
 A high-end 70% custom keyboard designed and produced by [jacky](https://geekhack.org/index.php?action=profile;u=63864) from Geekhack.
 
 Keyboard Maintainer: QMK Community
```
Comment (fauxpark) [inline keyboards/s7_elephant/readme.md:7]:
```suggestion
* Keyboard Maintainer: QMK Community
```

File: keyboards/s7_elephant/readme.md [hunk 10]
```diff
@@ -5,12 +5,17 @@
 A high-end 70% custom keyboard designed and produced by [jacky](https://geekhack.org/index.php?action=profile;u=63864) from Geekhack.
 
 Keyboard Maintainer: QMK Community  
-Hardware Supported: S7 Elephant  
-Hardware Availability: [The GB has ended with little chance to run again.](https://geekhack.org/index.php?topic=96434.0)  
+Hardware Supported: S7 Elephant rev1, rev2 
+Hardware Availability: [The GB has had 2 rounds with little chance to run again.](https://geekhack.org/index.php?topic=96434.0)
```
Comment (fauxpark) [inline keyboards/s7_elephant/readme.md:9]:
```suggestion
* Hardware Supported: S7 Elephant rev1, rev2
* Hardware Availability: [The GB has had 2 rounds with little chance to run again.](https://geekhack.org/index.php?topic=96434.0)
```

File: keyboards/s7_elephant/rev2/config.h [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+ * COLS: AVR pins used for columns, left to right
+ * ROWS: AVR pins used for rows, top to bottom
+ * DIODE_DIRECTION: COL2ROW = COL = Anode (+), ROW = Cathode (-, marked on diode)
+ *                  ROW2COL = ROW = Anode (+), COL = Cathode (-, marked on diode)
+ *
+ *
+ *                        0   1   2   3   4   5   6   7   8   9   A   B   C   D   E   F   */
+#define MATRIX_ROW_PINS { B0, B1, B2, B3, B7 }
+#define MATRIX_COL_PINS { D0, D1, D2, D3, D5, D4, D6, D7, B4, F7, F6, F5, F4, F1, F0, E6 }
+
```
Comment (mechmerlin) [inline keyboards/s7_elephant/rev2/config.h:41]: you should put the matrix size here, 5x16

File: keyboards/s7_elephant/rev2/info.json [hunk 12]
```diff
@@ -0,0 +1,90 @@
+{
+    "keyboard_name": "S7 Elephant",
+    "url": "https://geekhack.org/index.php?topic=103148.0",
+    "maintainer": "qmk",
+    "width": 17.25,
+    "height": 5,
+    "layouts": {
+        "LAYOUT": {
```
Comment (mechmerlin) [inline keyboards/s7_elephant/rev2/info.json:8]: you're missing the LAYOUT_all and LAYOUT_ansi which you defined in rev2.h

File: keyboards/s7_elephant/rev2/info.json [hunk 13]
```diff
@@ -0,0 +1,90 @@
+{
+    "keyboard_name": "S7 Elephant",
```
Comment (mechmerlin) [inline keyboards/s7_elephant/rev2/info.json:2]:
```suggestion
    "keyboard_name": "S7 Elephant Rev2",
```

File: keyboards/s7_elephant/rev2/rev2.h [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+ * ,------.   ,-------------------------------------------------------------------------------------------------.
+ * |      |   |       |     |     |     |     |     |     |     |     |     |     |     |     |           |     |
+ * |------|   |-------------------------------------------------------------------------------------------------|
+   |      |   |        |     |     |     |     |     |     |     |     |     |     |     |     |          |     |
+ * |------|   |-------------------------------------------------------------------------------------------------|
+ * |      |   |         |     |     |     |     |     |     |     |     |     |     |     |               |     |
+ * |------|   |-------------------------------------------------------------------------------------------------|
+ * |      |   |           |     |     |     |     |     |     |     |     |     |     |             |     |     |
+ * |------|   |-------------------------------------------------------------------------------------------------|
+ * |      |   |      |       |       |                                   |         |         |      |     |     |
```
Comment (mechmerlin) [inline keyboards/s7_elephant/rev2/rev2.h:43]: This ASCII art doesn't actually match your physical matrix defined in LAYOUT_ansi. Also LAYOUT_ansi doesn't match the LAYOUT_ansi of rev1. I would try to keep them the same.

File: keyboards/s7_elephant/rev2/info.json (hunk 12 above)
Comment (MudkipMao) [reply keyboards/s7_elephant/rev2/info.json:8]: What's the best way to generate an info.json? Adding those two layouts by hand doesn't sound like the correct methodology

Comment (mechmerlin) [reply keyboards/s7_elephant/rev2/info.json:8]: Noroadsleft tool only does it for the layout you created on kbfirmware. You can go about creating each layout again and just copy pasting the resulting LAYOUT macro, or you can start with the base .h file and modify it from there

Comment (noroadsleft) [reply keyboards/s7_elephant/rev2/info.json:8]: This actually matches the `LAYOUT_all` macro.

```suggestion
        "LAYOUT_all": {
```

Comment (MudkipMao) [reply keyboards/s7_elephant/rev2/info.json:8]: Yeah it does, I need to add the LAYOUT_ansi as well though, I'll get to that tonight

File: keyboards/s7_elephant/rev2/config.h (hunk 11 above)
Comment (MudkipMao) [reply keyboards/s7_elephant/rev2/config.h:41]: I have it in the config.h in the s7_elephant/ directory. However, I am going to refactor it to model after the kbd67 like you suggested last night

File: keyboards/s7_elephant/rev2/keymaps/default/keymap.c [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+ * |------|   |-------------------------------------------------------------------------------------------------|
+ * |  F3  |   | Caps    |  A  |  S  |  D  |  F  |  G  |  H  |  J  |  K  |  L  |  ;  |  '  |    Enter      | Home|
+ * |------|   |-------------------------------------------------------------------------------------------------|
+ * |  F4  |   | Shift     |  Z  |  X  |  C  |  V  |  B  |  N  |  M  |  ,  |  .  |  /  |    RShift   |  U  | End |
+ * |------|   |-------------------------------------------------------------------------------------------------|
+ * |  F5  |   | Ctrl |  Gui  |  Alt  |              Space                | RAlt |  FN  | RCtrl |  L  | D  |  R  |
+ * `------'   `-------------------------------------------------------------------------------------------------'
+ */
+
+    [_QWERTY] = LAYOUT_ansi(
```
Comment (noroadsleft) [inline keyboards/s7_elephant/rev2/keymaps/default/keymap.c:24]:
```suggestion
    [_QWERTY] = LAYOUT_ansi_1u(
```

File: keyboards/s7_elephant/rev2/keymaps/default/keymap.c [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+ * |------|   |-------------------------------------------------------------------------------------------------|
+ * |      |   |         |Prev |Play |Next |     |     |     |     |VolD |Mute |VolU |     |               |PgUp |
+ * |------|   |-------------------------------------------------------------------------------------------------|
+ * |      |   |           |     |     |     |     |     |     |     |     |     |     |             |     |PgDn |
+ * |------|   |-------------------------------------------------------------------------------------------------|
+ * |EEPRST|   |      |       |       |                                   |      |      |       |     |    |     |
+ * `------'   `-------------------------------------------------------------------------------------------------'
+ */
+
+    [_FN] = LAYOUT_ansi(
```
Comment (noroadsleft) [inline keyboards/s7_elephant/rev2/keymaps/default/keymap.c:46]:
```suggestion
    [_FN] = LAYOUT_ansi_1u(
```

File: keyboards/s7_elephant/rev2/keymaps/via/keymap.c [hunk 17]
```diff
@@ -0,0 +1,92 @@
...
+    * |  F2  |   | Tab    |  Q  |  W  |  E  |  R  |  T  |  Y  |  U  |  I  |  O  |  P  |  [  |  ]  |    \     | Del |
+    * |------|   |-------------------------------------------------------------------------------------------------|
+    * |  F3  |   | Caps    |  A  |  S  |  D  |  F  |  G  |  H  |  J  |  K  |  L  |  ;  |  '  |    Enter      |Home |
+    * |------|   |-------------------------------------------------------------------------------------------------|
+    * |  F4  |   | Shift     |  Z  |  X  |  C  |  V  |  B  |  N  |  M  |  ,  |  .  |  /  |    RShift   |  U  | End |
+    * |------|   |-------------------------------------------------------------------------------------------------|
+    * |  F5  |   | Ctrl |  Gui  |  Alt  |              Space                | RAlt |  FN  | RCtrl |  L  | D  |  R  |
+    * `------'   `-------------------------------------------------------------------------------------------------'
+    */
+    [_BASE] = LAYOUT_ansi(
```
Comment (noroadsleft) [inline keyboards/s7_elephant/rev2/keymaps/via/keymap.c:26]:
```suggestion
    [_BASE] = LAYOUT_ansi_1u(
```

File: keyboards/s7_elephant/rev2/keymaps/via/keymap.c [hunk 18]
```diff
@@ -0,0 +1,92 @@
...
+    * |  F7  |   |        |     |     |     |     |     |     |     |     |     |     |     |     |          |     |
+    * |------|   |-------------------------------------------------------------------------------------------------|
+    * |  F8  |   |         |VolD |VolU |Mute |     |     |     |     |     |     |     |     |               |PgUp |
+    * |------|   |-------------------------------------------------------------------------------------------------|
+    * |  F9  |   |           |Prev |Play |Next |     |     |     |     |     |     |     |             |     |PgDn |
+    * |------|   |-------------------------------------------------------------------------------------------------|
+    * |  F10 |   |      |       |       |                                   |      |      |       |     |    |     |
+    * `------'   `-------------------------------------------------------------------------------------------------'
+    */
+    [_FN1] = LAYOUT_ansi(
```
Comment (noroadsleft) [inline keyboards/s7_elephant/rev2/keymaps/via/keymap.c:46]:
```suggestion
    [_FN1] = LAYOUT_ansi_1u(
```

File: keyboards/s7_elephant/rev2/keymaps/via/keymap.c [hunk 19]
```diff
@@ -0,0 +1,92 @@
...
+    * |      |   |        |     |     |     |     |     |     |     |     |     |     |     |     |          |     |
+    * |------|   |-------------------------------------------------------------------------------------------------|
+    * |      |   |         |     |     |     |     |     |     |     |     |     |     |     |               |     |
+    * |------|   |-------------------------------------------------------------------------------------------------|
+    * |      |   |           |     |     |     |     |     |     |     |     |     |     |             |     |     |
+    * |------|   |-------------------------------------------------------------------------------------------------|
+    * |      |   |      |       |       |                                   |      |      |       |     |    |     |
+    * `------'   `-------------------------------------------------------------------------------------------------'
+    */
+    [_FN2] = LAYOUT_ansi(
```
Comment (noroadsleft) [inline keyboards/s7_elephant/rev2/keymaps/via/keymap.c:66]:
```suggestion
    [_FN2] = LAYOUT_ansi_1u(
```

File: keyboards/s7_elephant/rev2/keymaps/via/keymap.c [hunk 20]
```diff
@@ -0,0 +1,92 @@
...
+    * |      |   |        |     |     |     |     |     |     |     |     |     |     |     |     |          |     |
+    * |------|   |-------------------------------------------------------------------------------------------------|
+    * |      |   |         |     |     |     |     |     |     |     |     |     |     |     |               |     |
+    * |------|   |-------------------------------------------------------------------------------------------------|
+    * |      |   |           |     |     |     |     |     |     |     |     |     |     |             |     |     |
+    * |------|   |-------------------------------------------------------------------------------------------------|
+    * |      |   |      |       |       |                                   |      |      |       |     |    |     |
+    * `------'   `-------------------------------------------------------------------------------------------------'
+    */
+    [_FN3] = LAYOUT_ansi(
```
Comment (noroadsleft) [inline keyboards/s7_elephant/rev2/keymaps/via/keymap.c:86]:
```suggestion
    [_FN3] = LAYOUT_ansi_1u(
```

**Comment by MudkipMao** [comment]
> By the way, you can use the [Files Changed tab](https://github.com/qmk/qmk_firmware/pull/9277/files) to apply multiple suggestions to a commit.

Thanks.... That would be way better haha

**Comment by MudkipMao** [comment]
Ok, I pushed a new refactor, as well as took in the changes you and merlin have recommended. Both rev1 and rev2 compile. Lemme know anything looks off.

**Comment by MudkipMao** [comment]
Hey, I'm kinda new to this so what's different about the kb_breaking/ branch if it's going to get merged into master soon anyways? Does it run additional sanity checks or something?

**Comment by noroadsleft** [comment]
Much like the [Breaking Changes workflow](https://docs.qmk.fm/#/breaking_changes_instructions), this workflow is to only have breaking behavior occur at regular intervals, rather than on a random timeline. I'm pairing this with regular announcements on Reddit and Discord of the changes therein. Changes that relocated the code for single keyboards or multiple keyboards by one vendor used to go through the `future` branch (which is now named `develop`), which gets merged every three months. On the old workflow, these changes wouldn't arrive in QMK mainline until the last weekend of August.

I'm still working on getting some docs together to document the intended process.

**Comment by MudkipMao** [comment]
Thanks, that makes sense then! Last question, looks like travis is stuck right now. When I click on the build log, it says that it passed but here it says pending. Is there a way to manually retrigger a travis build?

**Comment by noroadsleft** [comment]
There is, but Travis is clear anyway. I think there's something wrong between Travis' notifications of build status and GitHub picking up the same. I get that issue even working on my own fork.

**Comment by MudkipMao** [comment]
So travis will build again on it's own?

**Comment by noroadsleft** [comment]
The Travis infrastructure only works against the `master` branch, unfortunately.

**Comment by MudkipMao** [comment]
Got it, so this needs to be merged manually by some before next Saturday right?

**Comment by noroadsleft** [comment]
@MudkipMao Can you rebase this branch against the `kb_breaking/2020-06-13` on QMK's repo? Do that and I'll merge this.

**Comment by MudkipMao** [comment]
I think I rebased this correctly... let me know if I didn't

**Comment by noroadsleft** [comment]
Ah, this isn't cleaner.

You'll technically lose the full change history, but the easiest way to clean this up is to run:

```shell
git checkout feature/s7_elephant_v2
# Make sure the command below references commit ec38f3f00ac7a0abf35817100a9da4b889419133
git log -1
git fetch upstream kb_breaking/2020-06-13                   # retrieves the state of QMK's kb_breaking/2020-06-13 branch
git reset --hard FETCH_HEAD                                 # rewrites your local files to match QMK's kb_breaking/2020-06-13 branch
cd keyboards/s7_elephant/~                                  
rm -r *                                                     # delete all the files in the s7_elephant directory; we'll get them back in a moment
git checkout ec38f3f00ac7a0abf35817100a9da4b889419133 -- .  # get the files we just deleted, but only for the current directory
git add -- .                                                # make sure all the files we need are staged to be committed
git commit -m "Add S7 Elephant Rev2 Support"                
git push --force-with-lease                                 # update your GitHub branch, overwriting the history
````

**Comment by MudkipMao** [comment]
Yeah, what I did definitely was not clean, I'll look at these commands later today. I'm very much a git noob sorry lol. I typically only use it for very basic workflows

**Comment by MudkipMao** [comment]
uh.... i followed your instructions and at the last step, I got the error:

fatal: You are not currently on a branch.
To push the history leading to the current (detached HEAD)
state now, use

    git push feature/s7_elephant_v2 HEAD:<name-of-remote-branch>



So I used this command:
git push --force-with-lease origin feature/s7_elephant_v2

not sure what that did..

**Pull request closed without merging**